### PR TITLE
[FEATURE][geocoder] Add a nominatim API geocoder

### PR DESCRIPTION
**Pull request by nirvn**
## Description

Title says it all. This PR adds a Nominatim API geocoder class. Not exposed through locator or as an algorithm (yet).

The geocoder is using the OSM-run nomimnatim endpoint by default. In order to respect their TOS, the nominatim geocoder class implements result caching and request throttling. :tada: 

@nyalldawson , review most appreciated.

### CODE REVIEW

File: src/core/geocoding/qgsnominatimgeocoder.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+     * \see setRequestsPerSecond()
+     */
+    int requestsPerSecond() const { return mRequestsPerSecond; }
+
+    /**
+     * Sets the \a number of request per seconds to the endpoint.
+     *
+     * \see requestsPerSecond()
+     */
+    void setRequestsPerSecond( int number ) { mRequestsPerSecond = number; }
```
Comment (nyalldawson) [inline src/core/geocoding/qgsnominatimgeocoder.h:92]: This should be a double, or the nicest you can be to the service is 1 request per second...

File: src/core/geocoding/qgsnominatimgeocoder.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Returns the number of requests per seconds to the endpoint.
+     *
+     * \see setRequestsPerSecond()
+     */
+    int requestsPerSecond() const { return mRequestsPerSecond; }
+
+    /**
+     * Sets the \a number of request per seconds to the endpoint.
+     *
```
Comment (nyalldawson) [inline src/core/geocoding/qgsnominatimgeocoder.h:89]: \warning Setting this to a value > 1 violates the nomatim terms of service. Only change this value if you are using a self-hosted nomatim service.

File: src/core/geocoding/qgsnominatimgeocoder.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+#include <QDateTime>
+#include <QUrl>
+#include <QUrlQuery>
+#include <QNetworkRequest>
+#include <QJsonDocument>
+#include <QJsonArray>
+
+QReadWriteLock QgsNominatimGeocoder::sMutex;
+QMap< QUrl, QList< QgsGeocoderResult > > QgsNominatimGeocoder::sCachedResults;
+qint64 QgsNominatimGeocoder::sLastRequestTimstamp = 0;
```
Comment (nyalldawson) [inline src/core/geocoding/qgsnominatimgeocoder.cpp:31]:
```suggestion
qint64 QgsNominatimGeocoder::sLastRequestTimestamp = 0;
```

File: src/core/geocoding/qgsnominatimgeocoder.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+  while ( QDateTime::currentMSecsSinceEpoch() - sLastRequestTimstamp < 1000 / mRequestsPerSecond )
+  {
+    QThread::msleep( 50 );
+    if ( feedback && feedback->isCanceled() )
+      return QList<QgsGeocoderResult>();
+  }
+  locker.changeMode( QgsReadWriteLocker::Write );
+  sLastRequestTimstamp = QDateTime::currentMSecsSinceEpoch();
+  locker.unlock();
```
Comment (nyalldawson) [inline src/core/geocoding/qgsnominatimgeocoder.cpp:106]: I would move this till after the blocking request call. Then you'd be excluding the length of time it takes to do the actual request from the "1 second between requests" logic and are being a bit more gentle on the service. By leaving the locker.changeMode to the Write mode before the request we ensure that only one request will be happening at a time.

File: src/core/geocoding/qgsnominatimgeocoder.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+  // Parse data
+  QJsonParseError err;
+  QJsonDocument doc = QJsonDocument::fromJson( newReq.reply().content(), &err );
+  if ( doc.isNull() )
+  {
+    return QList<QgsGeocoderResult>() << QgsGeocoderResult::errorResult( err.errorString() );
+  }
+
+  locker.changeMode( QgsReadWriteLocker::Write );
```
Comment (nyalldawson) [inline src/core/geocoding/qgsnominatimgeocoder.cpp:127]: Drop if above suggestion is applied -- we'd still have the write lock here already

File: src/core/geocoding/qgsnominatimgeocoder.cpp (hunk 4 above)
Comment (nirvn) [reply src/core/geocoding/qgsnominatimgeocoder.cpp:106]: @nyalldawson , the way I read Nominatum's usage policy, they set minimum to one request per second, not a 1 second gap between requests. If we move it below the blocking request is done, on very slow connections (i.e., developing parts of the world, where I code from :wink: ), you'd end up with say a request that lasts 1.5sec, then a 1sec gap for a total of 2.5sec.

Comment (nyalldawson) [reply src/core/geocoding/qgsnominatimgeocoder.cpp:106]: I don't personally see an issue with that, and think we're better to err on the side of being good citizens and keeping the load as light as possible...

**Review by nyalldawson — APPROVED** [review]
Looks good to me!

**Comment by nyalldawson** [comment]
Oh - one last thing -- don't we need to set a custom user agent too?

**Comment by nirvn** [comment]
@nyalldawson , our network access manager does it: https://github.com/qgis/QGIS/blob/master/src/core/qgsnetworkaccessmanager.cpp#L224

**Comment by nyalldawson** [comment]
Ok, great!

**Pull request merged**